### DISCUSSION TITLE
The clipboard on the sample client is not responding.

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8761] [Q&A]
Hello, I'm using Java to make freerdp available on the web. ( Not Android )

Previously, I implemented the clipboard by referring to the Android version, but I needed a file transfer function, 

so I referred to the wayland version.

Below is the init part of the clipboard that you have been using.
```
void tf_cliprdr_init(tfContext *tfc, CliprdrClientContext *cliprdr)
{

	cliprdr->custom = (void *)tfc;
	tfc->cliprdr = cliprdr;

	tfc->clipboard = ClipboardCreate();
	tfc->clipboardRequestEvent = CreateEvent(NULL, TRUE, FALSE, NULL);

	cliprdr->MonitorReady = tf_cliprdr_monitor_ready;
	cliprdr->ServerCapabilities = tf_cliprdr_server_capabilities;
	cliprdr->ServerFormatList = tf_cliprdr_server_format_list;
	cliprdr->ServerFormatListResponse = tf_cliprdr_server_format_list_response;
	cliprdr->ServerLockClipboardData = tf_cliprdr_server_lock_clipboard_data;
	cliprdr->ServerUnlockClipboardData = tf_cliprdr_server_unlock_clipboard_data;
	cliprdr->ServerFormatDataRequest = tf_cliprdr_server_format_data_request;
	cliprdr->ServerFormatDataResponse = tf_cliprdr_server_format_data_response;
	cliprdr->ServerFileContentsRequest = tf_cliprdr_server_file_contents_request;
	cliprdr->ServerFileContentsResponse = tf_cliprdr_server_file_contents_response;

```

Below is the init referring to wayland.

```
BOOL tf_cliprdr_init(tfClipboard* clipboard, CliprdrClientContext* cliprdr)
{

	if (!cliprdr || !clipboard)
		return FALSE;

	clipboard->context = cliprdr;
	cliprdr->custom = (void*)clipboard;
	cliprdr->MonitorReady = tf_cliprdr_monitor_ready;
	cliprdr->ServerCapabilities = tf_cliprdr_server_capabilities;
	cliprdr->ServerFormatList = tf_cliprdr_server_format_list;
	cliprdr->ServerFormatListResponse = tf_cliprdr_server_format_list_response;
	cliprdr->ServerFormatDataRequest = tf_cliprdr_server_format_data_request;
	cliprdr->ServerFormatDataResponse = tf_cliprdr_server_format_data_response;
	cliprdr->ServerFileContentsRequest = tf_cliprdr_server_file_contents_request;

	return TRUE;
}
```

By referring to the source, unlike the Android version, the clipboard structure was implemented in 'post_connect'. (clipboard_new)

There was no failure in 'init' and 'new', and no error log occurred.

After that, I ran the test and no response was received when copying.

Is there any part that needs to be set up differently from the Android version of the clipboard?

I also referred to the source of x11, but I don't know what's missing.

Please advise if there is anything that needs to be implemented after creating a clipboard structure using 'init' and 'new'. 

Thank you for reading it.

**Comment by akallabeth** [comment]
well, the clipboard is not a `I can just get something` but has a state machine that needs to be followed to get data.
See [MS-RDPECLIP](https://learn.microsoft.com/en-us/openspecs/windows_protocols/MS-RDPECLIP/fb9b7e0b-6db4-41c2-b83c-f889c1ee7688)

and of course, you need to implement the local backend  for your client (clipboard format announcements, and data request from local clipboard to initiate data transfer from server, ...)

**Reply by kimcoldwater** [reply]
https://woshub.com/cant-copy-paste-via-remote-desktop-rdp-clipboard/

It was this issue that there was no response from the clipboard.

What are the roles of `UwacClipboardOfferCreate` and `UwacClipboardOfferAnnounce` in `clipprdr_server_format_list` on the wayland version of the clipboard? 


Q1 . Is it possible to remove the part related to Uwac in the wayland version and implement it in the sample client?

I am checking and implementing X11 or Wayland versions, 

but we are having difficulty verifying the functionality of dependent libraries (Xlib, Uwac).



Q2 . can I implement a file clipboard on an Android version of the clipboard that doesn't include a dependent library?

I only need to be able to import file data. (No need for other information such as file path)

**Reply by kimcoldwater** [reply]
Missing `generalFlags` in `send_client_capabilities` was a problem. I looked at the Freerdp build option and the clipboard-related sources in the core and found that this part was missing. I didn't realize if this part was a problem because I set the build option for that part to ON. Anyway, it's settled.

`generalCapabilitySet.generalFlags |= CB_STREAM_FILECLIP_ENABLED | CB_FILECLIP_NO_FILE_PATHS | CB_HUGE_FILE_SUPPORT_ENABLED `
